walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,206 @@
+                            tempEnd = second.start();
+                            optional = true;
+                            tempStart = previous;
+                        }
+                    }
+                } else {
+                    if (optionalEvents.contains(first)) {
+                        if (first.start() - startTime < (int) request.getDuration()) {
+                            tempEnd = second.start();
+                            optional = true;
+                        }
+                    } else {
+                        tempEnd = first.start();
+                        previous = first.end();
+                    }
+                }
+
+                if (tempEnd - startTime >= (int) request.getDuration()) {
+                    times.add(TimeRange.fromStartEnd(startTime, tempEnd, false));
+                }
+
+                for (int i = 0; (i + 1) < sortedEvents.length; i++) {
+                    first = sortedEvents[i].getWhen();
+                    second = sortedEvents[i + 1].getWhen();
+
+                    if (first.overlaps(second)) {
+                        tempStart = Math.max(first.end(), second.end());
+
+                        if (optionalEvents.contains(first)) {
+                            if (first.end() - tempStart < (int) request.getDuration()) {
+                                tempStart = previous;
+                            }
+                        } else {
+                            previous = tempStart;
+                        }
+                    } else {
+                        tempStart = first.end();
+                        
+                        if (optionalEvents.contains(first)) {
+                            if (second.start() - first.end() < (int) request.getDuration()) {
+                                tempStart = previous;
+                            }
+                        } else {
+                            previous = tempStart;
+                        }
+                    }
+
+                    if (second.start() - tempStart >= (int) request.getDuration()) {
+                        times.add(TimeRange.fromStartEnd(tempStart, second.start(), false));
+                    }
+                }
+
+                tempStart = Math.max(previous, second.end());
+
+                if (endTime - tempStart >= (int) request.getDuration()) {
+                    times.add(TimeRange.fromStartEnd(tempStart, endTime, true));
+                }
+            
+                return times;
+            }
+        }
+    }
+
+    private static Event[] insertionSort(Event[] events, Comparator<TimeRange> comparator, MeetingRequest request) {
+        Event[] sortedArray = new Event[events.length];
+        ArrayList<Event> sorted = new ArrayList<>();
+        int back;
+
+        for (int k = 1; k < events.length; k++) {
+            back = k;
+            
+            while (back > 0) {
```

Comment:
This looks more like bubble sort than insertion sort.

Reply:
This is the insertion sort algorithm that was taught to me in one of my CS classes, could I ask why it instead looks like bubble sort to you?

Reply:
Your code conditionally swaps neighboring elements in increasing subsequences of the input.  That is exactly what bubble sorting does.  However, I realize that your version is a bit different than the standard version that stops when no more swapping is done.

https://en.wikipedia.org/wiki/Bubble_sort

The standard insertion sort involves finding the right place for an element and inserting it there, until no more elements remain.  That involves shifting the elements.  You could also imagine a variant that works by swapping instead of inserting, but that makes it "unstable", meaning the order of elements with identical keys can be different in the sorted order than the original list.

https://en.wikipedia.org/wiki/Insertion_sort

Since your code does conditional swapping it seemed more like bubble sort to me at first.  On closer inspection it does seem to behave more like insertion sort in that it grows a sorted list.  Regardless, both are equally inefficient :).

If you were writing this code I would suggest using nested for loops, which would make the code easier to understand (I also got misled by the fact that the inner loop was a while loop!):

```
for (int k = 1; k < events.length; ++k) {
  for (int back = k; back > 0; --back) {
     ...
  }
}
```

I also notice that you assign to sortedArray while you are sorting the events array in place.  I am not sure I understand the purpose of that.  In any case, like I had pointed out elsewhere, you would be better off using Arrays.sort to which you can supply a comparator.

Reply:
Ah I see, thank you for the insight and advice!

Reply:
Regarding using the Arrays sort method, I want to say that I actually have done so. I have pushed out another commit that has more comments , more clear variable names and code structure. Hopefully it will be easier to understand now!